### PR TITLE
Delegate and block producer clients.

### PR DESCRIPTION
**Pull request by romfrolov**
Summary:
- delegate client
- block producer client

### CODE REVIEW

File: core/account.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+ * General account @class.
+ */
+function Account(secretKey) {
+    if (!new.target) {
+        return new Account(secretKey);
+    }
+
+    while (!secretKey || !secp256k1.privateKeyVerify(secretKey)) {
+        secretKey = randomBytes(32);
+    }
```
Comment (damirka) [inline core/account.js:36]: Генерацию secret key я бы перенёс в отдельную функцию типа "generatePrivateKey". Кажется логичным.

File: core/blockchain.js [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+/**
+ * Handle stake transaction.
+ *
+ * @param  {String} tx
+ * @param  {Object} block
+ * @return {Object}
+ */
+function handleStake(tx, block) {
+    const sender = block.state.find(account => account.address === tx.from);
```
Comment (damirka) [inline core/blockchain.js]: Best practice - оборачивать boolean выражения в круглые скобки i.e. ``` => (account.address === tx.from)```
Отображает операцию и позволяет читать стрелочные функции как "действие => результат".

File: core/delegate.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    const finalRandomNumber = helpers.getRandomFromArray(randomNumbers);
+
+    if (isValidBlockProducer(block, finalRandomNumber) && isValidBlock(block)) {
+        transport.send({type: events.NEW_BLOCK, data: JSON.stringify(block)});
+    } else {
+        // in case when block is malformed we start round again
+        randomNumbers = [];
+    }
+
+    randomNumber = Math.random();
```
Comment (damirka) [inline core/delegate.js]: It may be better if didn't use decimals in random. Math.random() returns value in range 0-1.

Comment (damirka) [reply core/delegate.js]: So it's something like (Math.random() * 10).toFixed(0)

Comment (romfrolov) [reply core/delegate.js]: Sure. Random generation was just mocked out.

File: core/block-producer.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+ * @type {Number}
+ */
+const N_RANDOM_NUMBERS = 2;
+
+/**
+ * Delegates group name.
+ *
+ * @type {String}
+ */
+const DELEGATES_GROUP = 'delegates';
```
Comment (damirka) [inline core/block-producer.js]: = exports.DELEGATES_GROUP =

File: core/blockchain.js [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    return block;
+};
+
+/**
+ * Transaction handling entrypoint.
+ *
+ * @return {Object}
+ */
+exports.handleTransaction = function handleTransaction(tx, block) {
+    if (isVote(tx.data)) {
```
Comment (damirka) [inline core/blockchain.js]: switch (true) is so true.

File: core/blockchain.js [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+ *
+ * @param  {Object} block
+ * @param  {Number} finalRandomNumber
+ * @return {String}                   Address of block producer.
+ */
+function getBlockProducer(block, finalRandomNumber) {
+    let i = 0;
+
+    for (const account of block.state) {
+        for (let _ = 0; _ < account.certificates.length; _++) {
```
Comment (damirka) [inline core/blockchain.js]: Oi, gryazno. Nikakogo go-style, while in javascript.

**Review by damirka — APPROVED** [review]

File: core/blockchain.js (hunk 5 above)
Comment (romfrolov) [reply core/blockchain.js]: It just spawns a lot of `breaks`. Don't think it is worthy in this particular case.

Comment (damirka) [reply core/blockchain.js]: Instead of break there gonna be return statement which is even more clear.

File: core/blockchain.js [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+exports.getBlockProducer = getBlockProducer;
+
+/**
+ * Initial allocation of account balances from genesis.
+ *
+ * @param  {Object}  genesisBlock Genesis block.
+ * @param  {Object}  block
+ * @return {Object}
+ */
+exports.genesisAllocation = function genesisAllocation(genesisBlock, block) {
```
Comment (damirka) [inline core/blockchain.js]: Name functions with verbs. 
```
function allocateGenesis() {}
// ... or ...
function getGenesisBlock() {}
```

File: core/blockchain.js (hunk 6 above)
Comment (damirka) [reply core/blockchain.js]: It's i or any other letter. Underscore in JS means utility library.

File: core/delegate.js [hunk 8]
```diff
@@ -0,0 +1,95 @@
+/**
+ * Delegate client.
+ *
+ * @module core/delegate
```
Comment (damirka) [inline core/delegate.js]: Not a core. Move somewhere else. It's not a component - it's a nearly-complete client app.

File: core/delegate.js [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+// 4. Random number must be generated only once in a round.
+
+/**
+ * Delegates group name.
+ *
+ * @type {String}
+ */
+const DELEGATES_GROUP = 'delegates';
+
+const secretKey = '';
```
Comment (damirka) [inline core/delegate.js]: This is weird. Why?

**Review by damirka — CHANGES_REQUESTED** [review]

**Review by damirka — APPROVED** [review]
Yet it's not the best PR by code and architecture, it has to be merged for more comfrortable future development.

**Pull request merged**